### DISCUSSION TITLE
HOW TO  fastapi websocket  set up interface docker.APIClient exec_start

### DISCUSSION THREAD

**Discussion by binbinjianjian** [fastapi/fastapi#7269] [Questions]
I want to modify the flask_sockets  code written before through fastapi，but there is no idea

fastapi websocket The case cannot be imitated

import docker
import threading

class ClientHandler(object):

    def __init__(self, **kwargs):
        self.dockerClient = docker.APIClient(**kwargs)

    def creatTerminalExec(self, containerId):
        execCommand = [
            "/bin/sh",
            "-c",
            'TERM=xterm-256color; export TERM; [ -x /bin/bash ] && ([ -x /usr/bin/script ] && /usr/bin/script -q -c "/bin/bash" /dev/null || exec /bin/bash) || exec /bin/sh']
        execOptions = {
            "tty": True,
            "stdin": True
        }
        execId = self.dockerClient.exec_create(containerId, execCommand, **execOptions)
        print(execId)
        return execId["Id"]

    def startTerminalExec(self, execId):
        return self.dockerClient.exec_start(execId, socket=True, tty=True)

class DockerStreamThread(threading.Thread):
    def __init__(self, ws, terminalStream):
        super(DockerStreamThread, self).__init__()
        self.ws = ws
        self.terminalStream = terminalStream

    def run(self):
        while True:
            try:
                dockerStreamStdout = self.terminalStream.recv(2048)
                if dockerStreamStdout is not None:
                    self.ws.send(str(dockerStreamStdout, encoding='utf-8'))
                else:
                    print("docker daemon socket is close")
                    self.ws.close()
            except Exception as e:
                print("docker daemon socket err: %s" % e)
                self.ws.close()
                break

@sockets.route('/echo')
def echo_socket(ws):
    print(ws)
    dockerCli = ClientHandler(base_url=conf.DOCKER_HOST, timeout=600)
    terminalExecId = dockerCli.creatTerminalExec(conf.CONTAINER_ID)
    terminalStream = dockerCli.startTerminalExec(terminalExecId)._sock

    terminalThread = DockerStreamThread(ws, terminalStream)
    terminalThread.start()

    while not ws.closed:
        print('ok')
        message = ws.receive()
        if message is not None:
            terminalStream.send(bytes(message, encoding='utf-8'))

if __name__ == '__main__':
    from gevent import pywsgi
    from geventwebsocket.handler import WebSocketHandler

    app.debug = True
    server = pywsgi.WSGIServer(('192.168.100.1', 5000), app, handler_class=WebSocketHandler)
    server.serve_forever()

**Comment by tiangolo** [comment]
I assume you solved it, thanks for closing the issue :+1:
